src/backend/distributed/README.md

```diff
@@ -0,0 +1,2472 @@
+
+We do not precisely distinguish all possible cases, but rather have a simple distinction:
+
+- Fast path queries are simple queries on a single table with a <distribution column> = <Param or Const> filter (or single row inserts). We know that they prune to at most 1 shard regardless of the parameter value. The case of “distcol = NULL” is false/null by definition (unlike “distcol IS NULL”) and therefore prunes to 0 shards.
+- Router queries are arbitrarily complex queries that prune down to a single shard at planning time based on the RestrictInfo data structures obtained from postgres planner.
+
+We can only decide whether a query is a router query in the planner, because if it is not a router query, we need to fall back to the multi-shard query planning code path. Hence, we can only support generic router plans when all distribution column filters are constant, or there are only single shard/reference tables in the query. The router planner cannot prune based on unbound parameters and will therefore return a soft error. When the planner sees a soft error, we return a mock plan with a high cost, similar to multi-shard queries.
+
+Fast path queries prune to a single shard regardless of the parameter values. If the distribution column value is a parameter, we defer additional planning decisions, in particular “shard pruning” to the executor (deferredPruning flag in the Job). Currently, we resolve the parameters in ExecuteCoordinatorEvaluableExpressions which replaces the Param nodes in the query tree, and then TargetShardIntervalForFastPathQuery finds <distribution column> = <Const> filters in the WHERE clause. This could perhaps be optimized but keeps the logic consistent between parameters and non-parameterized queries.
+
+For both fast path queries and router queries, the job query tree for single shard queries still has all the parameters when we get to the executor. We resolve the parameters in the query tree before deparsing when:
+
+- pruning is deferred (has WHERE distcol = $1 …)
+- the query is a DML that contains function calls that need to resolved
+
+The latter happens primarily because function evaluation also resolves parameters. Otherwise, it could would not be able to resolve expressions like stable_fn($1). If the parameters are not resolved in the executor, they are passed on to the worker node using the libpq functions that take parameters.
+
+Both fast path and router query plans can be cached by PostgreSQL (plancache.c) if they are executed at least 5 times.
+
+- the query pruned to a single shard in the planner, the task is therefore static (always goes to the same shard group, with same query string)
```

Comment:
What do these two bullets belong to? I mean, which cases are they showing? The previous sentence doesn't seem complete and doesn't suggest there is a list coming up.
Previous sentence is:
`Both fast path and router query plans can be cached by PostgreSQL (plancache.c) if they are executed at least 5 times.`

Reply:
clarified further, the two cases are when distkey has a param vs no params